src/pykeen/nn/modules.py

```diff
@@ -311,6 +312,51 @@ def reset_parameters(self):
                 mod.reset_parameters()
 
 
+class LiteralInteraction(
+    Interaction,
+    Generic[HeadRepresentation, RelationRepresentation, TailRepresentation],
+):
+    """The interaction function shared by literal-containing interactions."""
+
+    def __init__(
+        self,
+        base: Interaction[HeadRepresentation, RelationRepresentation, TailRepresentation],
+        combination: nn.Module,
+    ):
+        # TODO documentation for this
+        super().__init__()
+        self.base = base
+        self.combination = combination
+        self.entity_shape = tuple(self.base.entity_shape) + ("e",)
+
+    def forward(
+        self,
+        h: HeadRepresentation,
+        r: RelationRepresentation,
+        t: TailRepresentation,
+    ) -> torch.FloatTensor:
+        """Compute broadcasted triple scores given broadcasted representations for head, relation and tails.
+
+        :param h: shape: (batch_size, num_heads, 1, 1, ``*``)
+            The head representations.
+        :param r: shape: (batch_size, 1, num_relations, 1, ``*``)
+            The relation representations.
+        :param t: shape: (batch_size, 1, 1, num_tails, ``*``)
+            The tail representations.
+
+        :return: shape: (batch_size, num_heads, num_relations, num_tails)
+            The scores.
+        """
+        # combine entity embeddings + literals
+        # TODO what if tail representation is not a single torch.FloatTensor?
+        h = torch.cat(h, dim=-1)
+        h = self.combination(h.view(-1, h.shape[-1])).view(*h.shape[:-1], -1)  # type: ignore
+        t = torch.cat(t, dim=-1)
+        # TODO what if tail representation is not a single torch.FloatTensor?
```

Comment:
Below this line `t` is a tensor. The actual question arises one line above: why do we always concat before?

Reply:
@mberr good question. can you look into this?

Reply:
I know the answer now. it's because h is a tuple of two tensors for the actual embedding then the literal embedding. Maybe it would be better to make the combination take them as two separate tensors and do its work on them rather than concatenating, then splitting apart again